homeassistant/components/repetier/__init__.py

```diff
@@ -0,0 +1,262 @@
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Repetier Server component."""
+    import pyrepetier
+
+    hass.data[REPETIER_API] = {}
+
+    for repetier in config[DOMAIN]:
+        _LOGGER.debug("Repetier server config %s", repetier[CONF_HOST])
+
+        url = "http://{}".format(repetier[CONF_HOST])
+        port = repetier[CONF_PORT]
+        api_key = repetier[CONF_API_KEY]
+
+        client = pyrepetier.Repetier(
+            url=url,
+            port=port,
+            apikey=api_key)
+
+        printers = client.getprinters()
+
+        if not printers:
+            return False
+
+        sensors = repetier[CONF_SENSORS][CONF_MONITORED_CONDITIONS]
+        api = PrinterAPI(hass, client, printers, sensors,
+                         repetier[CONF_NAME], config)
+        api.update()
+        track_time_interval(hass, api.update, SCAN_INTERVAL)
+
+        hass.data[REPETIER_API][repetier[CONF_NAME]] = api
+
+    return True
+
+
+class PrinterAPI:
+    """Handle the printer API."""
+
+    def __init__(self, hass, client, printers, sensors, conf_name, config):
+        """Set up instance."""
+        self._hass = hass
+        self._client = client
+        self.printers = printers
+        self.sensors = sensors
+        self.conf_name = conf_name
+        self.config = config
+        self._known_entities = []
+
+    def get_data(self, printer_id, sensor_type, temp_id):
+        """Get data from the state cache."""
+        printer = self.printers[printer_id]
+        methods = API_PRINTER_METHODS[sensor_type]
+        for prop, offline in methods['offline'].items():
+            state = getattr(printer, prop)
+            if state == offline:
+                # if state matches offline, sensor is offline
+                return None
+
+        data = {}
+        for prop, attr in methods['state'].items():
+            prop_data = getattr(printer, prop)
+            if attr == 'temp_data':
+                temp_methods = methods['temp_data']
+                for temp_prop, temp_attr in temp_methods.items():
+                    data[temp_attr] = getattr(prop_data[temp_id], temp_prop)
+            else:
+                data[attr] = prop_data
+        return data
+
+    def update(self, now=None):
+        """Update the state cache from the printer API."""
+        self._load_entities()
+        for printer in self.printers:
+            printer.get_data()
+        dispatcher_send(self._hass, UPDATE_SIGNAL)
+
+    def _load_entities(self):
+        sensor_info = []
+        for pidx, printer in enumerate(self.printers):
+            for sensor_type in self.sensors:
+                info = {}
+                info['sensor_type'] = sensor_type
+                info['printer_id'] = pidx
+                info['name'] = printer.slug
+                info['printer_name'] = self.conf_name
+                known = '{}-{}'.format(printer.slug, sensor_type)
+
+                if sensor_type == 'bed_temperature':
```

Comment:
Instead of checking sensor type, get the state dict of the sensor type from `API_PRINTER_METHODS` and check if `'temp_data'` is in the state dict values. If it is, we know that the sensor is a temperature sensor and needs to check for `None` in the return value of `getattr` on the printer using the key in the state dict as attribute, and an additional iteration of that attribute return value.

We can then limit the main cases of sensor types to temperature sensor and not temperature sensor.

Reply:
Took a few tests, but works now using `API_PRINTER_METHODS`